tests/VM_MessageBoxViewModel_Tests.cpp

```diff
@@ -0,0 +1,235 @@
+
+    class MockDesktop : public IDesktop, ra::services::ServiceLocator::ServiceOverride<IDesktop>
+    {
+    public:
+        MockDesktop(DialogHandler pHandler) : ra::services::ServiceLocator::ServiceOverride<IDesktop>(this), m_pHandler(pHandler)
+        {
+        }
+
+        ra::ui::DialogResult ShowModal(WindowViewModelBase& vmViewModel) const override
+        {
+            m_bDialogShown = true;
+
+            auto* vmMessageBox = dynamic_cast<MessageBoxViewModel*>(&vmViewModel);
+            if (vmMessageBox == nullptr)
+                return ra::ui::DialogResult::None;
+
+            return m_pHandler(*vmMessageBox);
+        }
+
+        void ShowWindow(WindowViewModelBase& vmViewModel) const override {}
+
+        void Shutdown() override {}
+
+        void AssertDialogShown()
+        {
+            Assert::IsTrue(m_bDialogShown);
+        }
+
+    private:
+        DialogHandler m_pHandler = nullptr;
+        mutable bool m_bDialogShown = false;
+    };
+
+    void AssertPointer(const WindowViewModelBase* pExpected, const WindowViewModelBase* pActual)
+    {
+        if (pExpected != pActual)
+        {
+            wchar_t buffer[128];
+            swprintf_s(buffer, 128, L"%p != %p", pExpected, pActual);
+            Assert::Fail(buffer);
+        }
+    }
+
+public:
+    TEST_METHOD(TestMessage)
+    {
+        MessageBoxViewModel vmMessageBox;
+        Assert::AreEqual(std::wstring(L""), vmMessageBox.GetMessage());
+
+        vmMessageBox.SetMessage(L"Test");
+        Assert::AreEqual(std::wstring(L"Test"), vmMessageBox.GetMessage());
+    }
+
+    TEST_METHOD(TestHeader)
+    {
+        MessageBoxViewModel vmMessageBox;
+        Assert::AreEqual(std::wstring(L""), vmMessageBox.GetHeader());
+
+        vmMessageBox.SetHeader(L"Test");
+        Assert::AreEqual(std::wstring(L"Test"), vmMessageBox.GetHeader());
+    }
+
+    TEST_METHOD(TestIcon)
+    {
+        MessageBoxViewModel vmMessageBox;
+        Assert::AreEqual(MessageBoxIcon::None, vmMessageBox.GetIcon());
+
+        vmMessageBox.SetIcon(MessageBoxIcon::Warning);
+        Assert::AreEqual(MessageBoxIcon::Warning, vmMessageBox.GetIcon());
+    }
+
+    TEST_METHOD(TestButtons)
+    {
+        MessageBoxViewModel vmMessageBox;
+        Assert::AreEqual(MessageBoxButtons::OK, vmMessageBox.GetButtons());
+
+        vmMessageBox.SetButtons(MessageBoxButtons::YesNo);
+        Assert::AreEqual(MessageBoxButtons::YesNo, vmMessageBox.GetButtons());
+    }
+
+    TEST_METHOD(TestShowMessage)
+    {
+        MockDesktop mockDesktop([](MessageBoxViewModel& vmMessageBox)
+        {
+            Assert::AreEqual(std::wstring(L"Hello, world!"), vmMessageBox.GetMessage());
+            Assert::AreEqual(std::wstring(), vmMessageBox.GetHeader());
+            Assert::AreEqual(MessageBoxIcon::None, vmMessageBox.GetIcon());
+            Assert::AreEqual(MessageBoxButtons::OK, vmMessageBox.GetButtons());
+            return DialogResult::OK;
+        });
```

Comment:
This illustrates how to use a MockDesktop helper class to intercept the request to display UI associated to a ViewModel and validate properties of the ViewModel. In the case of intercepting ShowModal, it returns the DialogResult.